File: go/euler14.go
package main
import "fmt"

var lengths = make(map[int]int)

func collatz(n int) int {
    if length, found := lengths[n]; found {
        return length
    }
    
    if n == 1 {
        return 1
    }
    
    var next int
    if n % 2 == 0 {
        next = n / 2
    } else {
        next = 3 * n + 1
    }
    
    length := 1 + collatz(next)
    lengths[n] = length
    return length
}

func main() {
    res := 0
    maxLen := 0
    lengths[1] = 1
    
    for i := 1; i < 1000000; i++ {
        length := collatz(i)
        if length > maxLen {
            res = i
            maxLen = length
        }
    }
    fmt.Println(res)
}

File: ts/euler14.ts
function longestCollatzSequence(): number {
    const cache: Map<number, number> = new Map();
    cache.set(1, 1);
    let res: number = 0, maxLen: number = 0;

    for (let i: number = 2; i < 1_000_000; i++) {
        let n: number = i, len: number = 0;
        while (n != 1 && !cache.has(n)) {
            if (n % 2 == 0) n /= 2;
            else n = 3 * n + 1;
            len++;
        }
        len += cache.get(n) ?? 0;
        cache.set(i, len);

        if (len > maxLen){
            maxLen = len;
            res = i;
        }
    }
    return res;
}

console.log(longestCollatzSequence());
